addons/website/tools.py

```diff
@@ -187,3 +191,149 @@ def get_base_domain(url, strip_www=False):
     if strip_www and url.startswith('www.'):
         url = url[4:]
     return url
+
+# Generated primary snippet templates
+
+def generate_primary_snippet_templates(env, module):
+    refs = {}
+
+    def ref(xmlid):
+        if xmlid not in refs:
+            refs[xmlid] = env.ref(xmlid).id
+        return refs[xmlid]
+
+    def split_key(snippet_key):
+        return snippet_key.split('.') if '.' in snippet_key else ('website', snippet_key)
+
+    def create_model_data(ids, values_list):
+        env['ir.model.data'].create([{
+            'name': values_list[index]['key'].split('.')[1],
+            'module': values_list[index]['key'].split('.')[0],
+            'model': 'ir.ui.view',
+            'res_id': ids[index],
+        } for index in range(0, len(ids))])
+        # Prevent deletion by _process_end
+        for values in values_list:
+            env['ir.model.data']._load_xmlid(values["key"])
+
+    def create_missing(values_list):
```

Comment:
They are separated because `get_create_vals` (formerly known as `generate_primary`) is called from 3 different nestings, and the various depths are used in the values (e.g. group, template name & snippet name) - but a single call to `create_missing` is needed for one given nesting depth.